src/reducers/computeRenderedColumns.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright Schrodinger, LLC
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule computeRenderedColumns
+ */
+
+'use strict';
+
+import clamp from 'lodash/clamp';
+import columnWidths from 'columnWidths';
+import scrollbarsVisibleSelector from 'scrollbarsVisible';
+import { updateColumnWidth, updateColumnGroupWidth } from 'updateColumnWidth';
+
+export default function computeRenderedColumns(state, columnAnchor) {
+  // clone state
+  const newState = Object.assign({}, state);
+  
+  // get buffer and viewport range
+  const columnRange = calculateRenderedColumnRange(newState, columnAnchor);
+  
+  // update the offsets and buffer mapping
+  computeRenderedColumnOffsets(newState, columnRange, newState.scrolling);
+
+  // scrollX might have changed due to change in columns and offsets
+  let scrollX = state.scrollX;
+  if (columnRange.firstViewportCol !== columnRange.endViewportCol) {
+    scrollX = newState.columnOffsets[columnRange.firstViewportCol] - newState.firstColumnOffset;
+  }
+  const { maxScrollX } = columnWidths(state);
+  newState.scrollX = clamp(scrollX, 0, maxScrollX);
+
+  return newState;
+}
+
+/**
+ * Determine the range of columns to render (buffer and viewport)
+ * The leading and trailing buffer is based on a fixed count,
+ * while the viewport columns are based on their width and the viewport width.
+ * We use the columnAnchor to determine what either the first or last column
+ * will be, as well as the offset.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ * }} columnAnchor
+ * @return {{
+ *   endBufferCol: number,
+ *   endViewportCol: number,
+ *   firstBufferCol: number,
+ *   firstViewportCol: number,
+ * }}
+ * @private
+ */
+function calculateRenderedColumnRange(state, columnAnchor) {
+  const bufferColumnCount = 0; // TODO (pradeep): calculate this similar to bufferRowCount
+  const { availableScrollWidth, scrollableColumns } = columnWidths(state);
+  const columnCount = scrollableColumns.length;
+
+  if (availableScrollWidth === 0 || columnCount === 0) {
+    state.firstColumnIndex = 0;
+    state.endColumnIndex = 0;
+    state.firstColumnOffset = 0;
+    return {
+      endBufferCol: 0,
+      endViewportCol: 0,
+      firstBufferCol: 0,
+      firstViewportCol: 0,
+    };
+  }
+
+  // If our first or last index is greater than our columnCount,
+  // treat it as if the last column is at the end of the viewport
+  let { firstIndex, firstOffset, lastIndex } = columnAnchor;
+  if (firstIndex >= columnCount || lastIndex >= columnCount) {
+    lastIndex = columnCount - 1;
```

Comment:
lastIndex may also be a column earlier in the table.  

A case where that could happen would be if I have a table with 50 columns where 20 are off the screen to the right.  If I make a scrollToColumn call on column on the 45th column, the table will scroll so that the 45th column appears as the rightmost column in the table.

In that case lastIndex will be set to 45 so the 45th column is shown as the last column in the scrollable section.  (apologies for any off by one errors in my description)

Did that make any sense?  If not I can clarify on a whiteboard.